### PR TITLE
Integrate attention sink into ET LLM export and runner (#18860)

### PR DESCRIPTION
**Pull request by kirklandsign**
Summary:

Add custom op support, export pipeline integration, and C++ runner fixes
for the attention sink ring buffer implementation.

- CustomKVCacheWithAttentionSink: custom op variant using update_cache_with_indices
  for scatter-write performance. Replaces KVCacheWithAttentionSink during export.
- CustomRingKVCache replacement: handle RingKVCache -> CustomRingKVCache in the
  replacement pass, and set SDPACustom.use_attention_mask=True for ring buffer models.
- Export transform ordering: replace SDPA before KV cache so that
  _replace_kv_cache_with_custom_kv_cache can set use_attention_mask=True on the
  already-existing SDPACustom (previously the ordering was reversed, causing the
  mask flag to be overwritten by a new SDPACustom).
- C++ runner: add max_seq_len prefill check; make context length check conditional
  for sliding window models (max_seq_len < max_context_len) since they handle
  position wrapping internally via ring buffer.

Reviewed By: lucylq

Differential Revision: D100216686

### CODE REVIEW

**Comment by kirklandsign** [comment]
@claude review the epatch

**Comment by kirklandsign** [comment]
> **Claude finished @kirklandsign's task in 1m 51s** —— [View job](https://github.com/pytorch/executorch/actions/runs/24373744470)
> 
> ### Review of PR #18860: Integrate attention sink into ET LLM export and runner
> * [x]  Read the full diff and changed files
> * [x]  Review `custom_kv_cache.py` changes
> * [x]  Review `export_llama_lib.py` changes
> * [x]  Review `text_llm_runner.cpp` changes
> * [x]  Review `test_attention_sink.py` changes
> * [x]  Post review feedback
> 
> Overall this is a well-structured PR that correctly integrates the attention sink ring buffer into the export pipeline and C++ runner. The transform ordering fix is sound, the new `CustomKVCacheWithAttentionSink` class mirrors existing patterns, and the tests cover the key scenarios. A few items worth addressing:
> 
> ### Issue 1: Missing `num_window_tokens <= ring_size` guard in `CustomKVCacheWithAttentionSink.update()`
> `KVCacheWithAttentionSink.update()` (`attention_sink.py:240-246`) has an explicit guard that the number of window tokens doesn't exceed the ring buffer capacity — this prevents duplicate indices in the scatter write, which would cause non-deterministic behavior:
> 
> ```python
> # attention_sink.py:239-245
> start_pos = input_pos[0].item()
> num_sink_tokens = max(0, min(seq_len, self.sink_size - start_pos))
> num_window_tokens = seq_len - num_sink_tokens
> assert num_window_tokens <= self.cache_positions_manager.ring_size, (
>     ...
> )
> ```
> 
> `CustomKVCacheWithAttentionSink.update()` (`custom_kv_cache.py:551-561`) uses `update_cache_with_indices` (also a scatter-style operation) but lacks this guard. While the existing `seq_len <= cache_size` check at line 553 prevents writing more tokens than total cache slots, it doesn't prevent the ring portion from wrapping with duplicate indices when `num_window_tokens > ring_size`.
> 
> Consider adding the same guard before computing indices.
> 
> [Fix this →](https://claude.ai/code?q=In%20examples%2Fmodels%2Fllama%2Fsource_transformation%2Fcustom_kv_cache.py%2C%20add%20the%20same%20num_window_tokens%20%3C%3D%20ring_size%20guard%20that%20exists%20in%20KVCacheWithAttentionSink.update()%20(attention_sink.py%3A239-245)%20to%20CustomKVCacheWithAttentionSink.update()%20around%20line%20551-561.%20Branch%3A%20export-D100216686&repo=pytorch/executorch)
> 
> ### Issue 2: `effective_pos = 0` may over-allocate generation budget
> In `text_llm_runner.cpp:188`:
> 
> ```c++
> int64_t effective_pos = (max_seq_len < max_context_len) ? 0 : pos_;
> int max_new_tokens = config.resolve_max_new_tokens(max_context_len, effective_pos);
> ```
> 
> For sliding window models, passing `effective_pos = 0` means `resolve_max_new_tokens` returns `max_context_len` (or close to it) as the generation budget. This is correct in the sense that the ring buffer can recycle space indefinitely, so the KV cache won't overflow. However, this effectively means generation can proceed until `pos_` exceeds `max_context_len`, which is the RoPE frequency table size. If the model's RoPE table was exported with a specific `max_context_len` bound, generating past it would cause an out-of-bounds access on the frequency table.
> 
> In practice this may be fine if the caller sets `max_new_tokens` explicitly, but without that, the runner silently allows generating up to `max_context_len` tokens regardless of current position. Consider whether this should be documented more explicitly or whether the RoPE table size should be the real cap (it already is, but the connection is implicit). This is a minor concern since the RoPE module itself will guard against exceeding its table size.
> 
> ### Issue 3: `_replace_kv_cache_with_custom_kv_cache` doesn't recurse for matched cases
> In `custom_kv_cache.py:379-424`, when a child matches `KVCacheWithAttentionSink`, `RingKVCache`, or `KVCache`, the function replaces it but does **not** recurse into the replacement or the child's own children. The `else` branch at line 423 is only reached if none of the `isinstance` checks match.
> 
> This is fine for the current model structure where KV caches are direct children of attention modules, but it means nested structures won't be traversed if a match is found at a parent level. The existing code before this PR had the same pattern for `KVCache`, so this is consistent — just worth noting that it assumes flat module hierarchies for cache modules.
> 
> ### Issue 4: Storing a module-level function reference as an instance attribute
> `custom_kv_cache.py:536-538`:
> 
> ```python
> self._create_causal_mask_for_attention_sink = (
>     _create_causal_mask_for_attention_sink
> )
> ```
> 
> This stores a module-level function as an instance attribute. While functional, it's unusual and could cause issues with `torch.export` / `torch.jit` tracing since it's not a submodule or buffer. `CustomRingKVCache` (line 594) avoids this by calling `_create_causal_mask_for_ring_buffer` directly from module scope. Consider doing the same for `_create_causal_mask_for_attention_sink` — importing at the top of the class method rather than storing as an attribute.
> 
> ### Minor observations
> * **Transform ordering fix** (`export_llama_lib.py`): The reordering from KV-cache-then-SDPA to SDPA-then-KV-cache is correct. The comment at lines 1784-1787 explains the rationale clearly.
> * **Test coverage** (`test_attention_sink.py:401-444`): The two new tests (`test_beyond_context_window_custom_sdpa` and `test_sink_zero_custom_sdpa`) cover the main paths well. The test at line 348 correctly calls `replace_sdpa_with_custom_op` before `replace_kv_cache_with_custom_kv_cache`, matching the new transform order.
> * **Error message improvement** (`text_llm_runner.cpp:143-148`): The new prefill check against `max_seq_len` gives a better error message than the old combined check. The separation of prefill-chunk-size vs. context-capacity checks is an improvement.

Issue 1 and 4 will be addressed in a follow-up patch. 2 and 3 is not a blocker

**Review by lucylq — APPROVED** [review]
